core/chains/evm/logpoller/log_poller.go

```diff
@@ -707,23 +706,62 @@ func (lp *logPoller) LatestLogEventSigsAddrsWithConfs(fromBlock int64, eventSigs
 	return lp.orm.SelectLatestLogEventSigsAddrsWithConfs(fromBlock, addresses, eventSigs, confs, qopts...)
 }
 
-// GetBlocks tries to get the specified block numbers from the log pollers
-// blocks table. Returns the blocks it was able to find, empty slice if none.
-// When the log poller does not have requested blocks, it falls back
-// to RPC to fetch the missing blocks.
-// response contains blocks in the same order as "numbers" in request parameters
-// the first context parameter takes precedence over contexts passed through qopts
-func (lp *logPoller) GetBlocks(ctx context.Context, numbers []uint64, qopts ...pg.QOpt) ([]LogPollerBlock, error) {
+// GetBlocksAscendingUsingRange tries to get the specified block numbers from the log pollers
+// blocks table, and sorts them in ascending order. It falls back to the RPC for any unfulfilled
+// blocks.
+//
+// Note: it does not fetch any unrequested blocks from the RPC.
+func (lp *logPoller) GetBlocksAscendingUsingRange(ctx context.Context, numbers []uint64, qopts ...pg.QOpt) ([]LogPollerBlock, error) {
+	var blocks []LogPollerBlock
+
+	// Do nothing if no blocks are requested.
+	if len(numbers) == 0 {
+		return blocks, nil
+	}
+
+	// Sort the requested blocks in ascending order and assign them to a request map.
+	sort.Slice(numbers, func(i, j int) bool { return numbers[i] < numbers[j] })
+	blocksRequested := make(map[uint64]struct{})
+	for _, b := range numbers {
+		blocksRequested[b] = struct{}{}
+	}
+
+	// Retrieve all blocks within this range from the log poller.
 	blocksFound := make(map[uint64]LogPollerBlock)
 	qopts = append(qopts, pg.WithParentCtx(ctx))
-	lpBlocks, err := lp.orm.GetBlocks(numbers, qopts...)
+	lpBlocks, err := lp.orm.GetBlocksFromRange(numbers[0], numbers[len(numbers)-1], qopts...)
 	if err != nil {
 		lp.lggr.Warnw("Error while retrieving blocks from log pollers blocks table. Falling back to RPC...", "requestedBlocks", numbers, "err", err)
 	}
 	for _, b := range lpBlocks {
-		blocksFound[uint64(b.BlockNumber)] = b
+		if _, ok := blocksRequested[uint64(b.BlockNumber)]; ok {
+			// Only fill requested blocks.
+			blocksFound[uint64(b.BlockNumber)] = b
+		}
+	}
+
+	// Fill any remaining blocks from the client.
+	err = lp.fillRemainingBlocksFromRPC(ctx, blocksFound, numbers)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, num := range numbers {
+		b, ok := blocksFound[num]
+		if !ok {
+			return nil, errors.Errorf("block: %d was not found in db or RPC call", num)
+		}
+		blocks = append(blocks, b)
 	}
 
+	return blocks, nil
+}
+
+func (lp *logPoller) fillRemainingBlocksFromRPC(
+	ctx context.Context,
+	blocksFound map[uint64]LogPollerBlock,
```

Comment:
I think it's cleaner not to have these kinds of output parameters, pure functions are easier to grok than impure ones like this one. You can merge the output of this with the map you have in `GetBlocksAscendingUsingRange` pretty easily since it's a `map`.